File: 25. Reverse Nodes in k-Group.py
# https://leetcode.com/problems/reverse-nodes-in-k-group/

# Given a linked list, reverse the nodes of a linked list k at a time and return
# its modified list.

# k is a positive integer and is less than or equal to the length of the linked
# list. If the number of nodes is not a multiple of k then left-out nodes, in the
# end, should remain as it is.

# You may not alter the values in the list's nodes, only nodes themselves may be
# changed.

################################################################################

# reverse every sub-ListNode of length k
# curr = dummy, move k steps to make curr = tail
# prev -> (head -> ... -> tail) -> next
# prev -> (tail -> ... -> head) -> next
# separate func to reverse (head, tail)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverse(self, head, tail):
        prev, curr = None, head
        while prev != tail:
            next = curr.next # (prev -> curr) -> next
            curr.next = prev # (prev <- curr)    next
            
            prev, curr = curr, next # move on
        return tail, head

    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy = ListNode()
        dummy.next = head
        prev = curr = dummy

        while curr.next:
            for i in range(k): # move curr k steps to make curr = tail
                curr = curr.next
                if not curr: # curr = None, not tail
                    return dummy.next
            
            # prev -> (head -> ... -> tail) -> next
            head = prev.next
            tail = curr
            next = curr.next
            
            tail, head = self.reverse(head, tail)
            
            # reconnect: prev -> (tail -> ... -> head) -> next
            prev.next = tail
            head.next = next

            # move prev and curr
            prev = curr = head
        return dummy.next

File: 25. Reverse Nodes in k-Group.cpp
// https://leetcode.com/problems/reverse-nodes-in-k-group/

// Given a linked list, reverse the nodes of a linked list k at a time and return
// its modified list.

// k is a positive integer and is less than or equal to the length of the linked
// list. If the number of nodes is not a multiple of k then left-out nodes, in the
// end, should remain as it is.

// You may not alter the values in the list's nodes, only nodes themselves may be
// changed.

////////////////////////////////////////////////////////////////////////////////

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(); dummy->next = head;
        ListNode* ptrCurr = dummy;
        ListNode* ptrHeadPrev = dummy;
        while (ptrCurr->next) {
            // move ptrCurr k times to thisTail
            for (int i = 0; i < k; ++i) {
                ptrCurr = ptrCurr->next;
                if (ptrCurr == nullptr) return dummy->next;
            }
            // ptrHeadPrev -> (ptrHead -> ... -> ptrTail) -> ptrTailNext
            ListNode* ptrHead = ptrHeadPrev->next;
            ListNode* ptrTail = ptrCurr;
            ListNode* ptrTailNext = ptrCurr->next;
            // ptrHeadPrev -> (ptrTail -> ... -> ptrHead) -> ptrTailNext            
            tie(ptrTail, ptrHead) = reverse(ptrHead, ptrTail);
            ptrHeadPrev->next = ptrTail;
            ptrHead->next = ptrTailNext;
            ptrHeadPrev = ptrHead;
            ptrCurr = ptrHead;
        }
        return dummy->next;
    }
    
    pair<ListNode*, ListNode*> reverse(ListNode* ptrHead, ListNode* ptrTail) {
        ListNode* ptrPrev = new ListNode();
        ListNode* ptrCurr = ptrHead;
        // reverse prev and curr
        while (ptrPrev != ptrTail) {
            ListNode *ptrNext = ptrCurr->next; // prev -> curr -> next
            ptrCurr->next = ptrPrev; // prev <- curr    next
            ptrPrev = ptrCurr;
            ptrCurr = ptrNext;
        }
        return make_pair(ptrTail, ptrHead);
    }
};
